### ISSUE TITLE
feat(latents): re-enable effective_einstein_radius via zero_contour

### ISSUE THREAD

**Issue by Jammy2211** [PyAutoLabs/euclid_strong_lens_modeling_pipeline#14]
## Overview

Re-enables the `effective_einstein_radius` latent variable in `util.py` (currently commented out at line 491) by routing it through the JAX-traceable `einstein_radius_via_zero_contour_from()` for JAX runs and keeping the legacy marching-squares `einstein_radius_from(grid=...)` for `use_jax=False` runs. Phase B of [`z_features/fast_visualization.md`](https://github.com/PyAutoLabs/PyAutoPrompt/blob/main/z_features/fast_visualization.md), unblocked by the perf cache fix in PyAutoGalaxy #434 (warm calls ~10 s → ~68 ms).

## Plan

- Uncomment + rename the `latent_effective_einstein_radius` entry in `LATENT_KEYS` to `latent.effective_einstein_radius` so the dotted naming matches all the other keys.
- Replace the commented-out compute block with a dispatch that branches on `self._use_jax`: zero_contour for JAX runs, marching squares for numpy runs. `except ValueError` only — degenerate models (no zero crossing) → NaN; anything else propagates.
- Uncomment the `effective_einstein_radius` slot in the returned tuple so its length matches `LATENT_KEYS` (now 9).
- Smoke-test both backends: `PYAUTO_DISABLE_JAX=1` (numpy branch) via the standard smoke list; one manual run without `PYAUTO_DISABLE_JAX` (JAX branch) to confirm the JAX-traceable path lands the latent.
- No `z_projects/euclid` edits, no library changes, no extension of `LATENT_KEYS` beyond Einstein radius.

<details>
<summary>Detailed implementation plan</summary>

### Affected Repositories
- euclid_strong_lens_modeling_pipeline (primary)

### Work Classification
Workspace.

### Branch Survey

| Repository | Current Branch | Dirty? |
|---|---|---|
| ./euclid_strong_lens_modeling_pipeline | main | clean |

**Suggested branch:** `feature/euclid-einstein-radius-zero-contour`
**Worktree root:** `~/Code/PyAutoLabs-wt/euclid-einstein-radius-zero-contour/` (created later by `/start_workspace`)

### Implementation Steps

1. **`util.py` `LATENT_KEYS` (line 305–315)** — uncomment the `latent_effective_einstein_radius` entry and rename to `latent.effective_einstein_radius` for consistency with the dotted format. Final list length: 9.
2. **`util.py` `compute_latent_variables` Einstein-radius block (line 488–495)** — replace the commented-out block with the conditional dispatch:
   ```python
   try:
       if self._use_jax:
           effective_einstein_radius = tracer.einstein_radius_via_zero_contour_from()
       else:
           effective_einstein_radius = tracer.einstein_radius_from(
               grid=self.dataset.grids.lp,
           )
   except ValueError:
       effective_einstein_radius = xp.nan
   ```
3. **`util.py` returned tuple (line 497–507)** — uncomment the `effective_einstein_radius` line so the tuple length matches `LATENT_KEYS` length (9).
4. **Smoke + manual verification** — `/smoke_test euclid_strong_lens_modeling_pipeline` exercises the numpy branch under `PYAUTO_DISABLE_JAX=1`. One manual `python start_here.py --dataset=... --sample=...` run without `PYAUTO_DISABLE_JAX` exercises the JAX branch; inspect `output/.../latent/latent_summary.json` for finite `latent.effective_einstein_radius`.

### Key Files
- `util.py:305-315` — `LATENT_KEYS` rename + uncomment
- `util.py:488-507` — conditional dispatch + tuple update

### Out of scope
- No `z_projects/euclid` edits (live science work for the DR1 catalogue paper).
- No library changes — the API (`einstein_radius_via_zero_contour_from()`) already exists in PyAutoGalaxy.
- No extension of `LATENT_KEYS` beyond Einstein radius — additional latents (critical-curve area, position-magnification, etc.) are separate prompts.

</details>

## Original Prompt

<details>
<summary>Click to expand starting prompt</summary>

See `PyAutoPrompt/issued/einstein_radius_zero_contour_migration.md` for the verbatim authored prompt. Key reference points kept inline:

- **The latent has been disabled** since `tracer.einstein_radius_from(grid=...)` is not JAX-traceable. `compute_latent_variables` runs once per posterior sample inside the JAX vmap, so the legacy method either had to force `self._use_jax = False` per sample (the workaround used in `z_projects/euclid`) or skip the latent entirely (this pipeline's approach until now).
- **PyAutoGalaxy PR #434** (merged 2026-05-21) cached `(f, ZeroSolver)` on `LensCalc`. Warm-call latency on a typical SIE: tangential CC 67 ms, Einstein radius 68 ms (CPU). Without the cache, warm calls were ~10 s — unusable per-sample.
- **The dispatch on `self._use_jax`** is the same pattern revised Phase A applies at the library level: JAX callers route to zero_contour (traceable, fast warm), numpy callers stay on marching squares (no JAX dep, fast cold).
- **`except ValueError` only.** `_init_guess_from_coarse_grid` raises `ValueError` for degenerate models with no zero crossings — return NaN for that sample. Other exceptions propagate loudly per `feedback_no_silent_guards`.

</details>

**Comment by Jammy2211** [comment]
## PRs Created — 2026-05-21

### Library PRs
- PyAutoFit: https://github.com/PyAutoLabs/PyAutoFit/pull/1288 — `Analysis.LATENT_BATCH_MODE` attribute (vmap default, jit option). CI: all green (1401 tests passed locally, all 6 CI checks green).
- PyAutoGalaxy: https://github.com/PyAutoLabs/PyAutoGalaxy/pull/435 — `LensCalc.einstein_radius_jit_from` helper + `AnalysisDataset.LATENT_BATCH_MODE = "jit"`. CI: all green (926 tests passed locally, all 5 unittest CI checks green).

### Workspace PR
- This repo: https://github.com/PyAutoLabs/euclid_strong_lens_modeling_pipeline/pull/15 — re-enables `latent.effective_einstein_radius` via the new helper.

### Status
Library PRs are green and ready to merge. Workspace PR's `url_check` CI is failing on a **pre-existing** `Jammy2211/autolens_workspace` reference in `README.md:110` (line 110, not touched by this PR). That predates this work and is unrelated.

### How the design landed

`compute_latent_samples` cannot use `jax.vmap` for lensing latents because the underlying `jax_zero_contour.ZeroSolver.zero_contour_finder` uses `jax.lax.cond` and `jax.lax.while_loop` for early termination, which upstream explicitly documents as incompatible with `jax.vmap`. The path that landed:

1. **PyAutoFit** adds a `LATENT_BATCH_MODE` switch — default stays `"vmap"` so non-lensing users are unchanged.
2. **PyAutoGalaxy** sets `AnalysisDataset.LATENT_BATCH_MODE = "jit"` and adds `LensCalc.einstein_radius_jit_from(init_guess, ...)`. The new helper skips the un-jit-able pieces (skimage seed search, `path_reduce`, Python `for`-loop) and computes shoelace area on the raw NaN-padded paths array.
3. **Workspace** dispatches `if self._use_jax` to the new helper with a 4-seed fan at ±1 arcsec, else to legacy `einstein_radius_from(grid=...)`.

### Verification
- PyAutoFit `pytest test_autofit/analysis/`: 18 passed, 1 skipped.
- PyAutoGalaxy `pytest test_autogalaxy/operate/ test_autogalaxy/analysis/`: 78 passed.
- Pipeline numpy-branch smoke (`PYAUTO_DISABLE_JAX=1`): 6/6 PASS.
- Pipeline JAX-branch end-to-end (`start_here.py`, `PYAUTO_TEST_MODE=1`, JAX enabled): max-LL `latent.effective_einstein_radius = 2.1002 arcsec`, ~480 ms/sample after the ~10 s ZeroSolver compile.

### Merge gate
Library PRs (#1288, #435) merge first → release → workspace PR (#15) can merge.

**Comment by Jammy2211** [comment]
## Shipped — 2026-05-21

### PRs
- PyAutoFit: https://github.com/PyAutoLabs/PyAutoFit/pull/1288 (merged fe8f6d94)
- PyAutoGalaxy: https://github.com/PyAutoLabs/PyAutoGalaxy/pull/435 (merged 2547ca17)
- Pipeline workspace: https://github.com/PyAutoLabs/euclid_strong_lens_modeling_pipeline/pull/15 (merged 5fd0a550)

### Summary

`latent.effective_einstein_radius` is back in the Euclid pipeline. Architecture: PyAutoFit's `Analysis.LATENT_BATCH_MODE` switch (default `"vmap"`, opt-in `"jit"`) + PyAutoGalaxy's `LensCalc.einstein_radius_jit_from` (jit-friendly helper that bypasses skimage seed search and `path_reduce`) + `AnalysisDataset.LATENT_BATCH_MODE = "jit"` so lensing analyses use per-sample jit automatically. Verified end-to-end: max-LL `latent.effective_einstein_radius = 2.1002` arcsec on the prior-median MGE tracer.

### Session Notes

- The original Phase B premise (drop-in swap to `_via_zero_contour_from()` inside `compute_latent_samples`' vmap) was invalid — `jax_zero_contour.ZeroSolver` upstream documents that its early-termination `lax.cond` / `lax.while_loop` is **not safe under vmap**. The chosen path honours that contract: per-sample `jax.jit` (not `jit(vmap(...))`), reuses the compile cache across samples without hitting the vmap perf penalty.
- The original prompt also told the developer NOT to flip `_use_jax = False` for latents. That guidance stands — the new helper *is* JAX-traceable, the dispatch just branches on `self._use_jax` at the workspace level rather than forcing a runtime flip.
- Speed: ~480 ms/sample after the ~10 s first-sample compile. For Euclid scale, the per-sample numpy workaround (`z_projects/euclid` pattern) is actually faster (~30 ms/sample, ~30 s for 1000 samples vs ~80 s here). The jit-only path was chosen for architectural consistency and because it scales better for cluster-scale geometry where marching squares becomes a bottleneck.
- PR body URL cross-references hit a `gh pr edit` GraphQL deprecation warning; the subagent patched via `gh api PATCH` as a workaround (same as the previous task — worth fixing in the gh upgrade, separate concern).
- Workspace PR merged with a red `url_check` CI — pre-existing `Jammy2211/autolens_workspace` reference in `README.md:110`, unrelated to this PR. Library PRs were both fully green.
